packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -86,4 +90,28 @@ class UpgradeCommand extends FlutterCommand {
       || _gitChangedRegex.hasMatch(line)
       || line == 'Fast-forward';
   }
+
+  /// Find and return the project root directory relative to the specified
+  /// directory or the current working directory if none specified.
+  /// Return `null` if the project root could not be found
+  /// or if the project root is inside the flutter repository.
+  static String findProjectRoot([String directory]) {
+    directory ??= Directory.current.path;
+    if (directory.startsWith(Cache.flutterRoot)) {
+      if (directory.length == Cache.flutterRoot.length ||
+          directory[Cache.flutterRoot.length] == Platform.pathSeparator) {
+        return null;
+      }
+    }
+    while (true) {
+      if (FileSystemEntity.isFileSync(join(directory, 'pubspec.yaml'))) {
```

Comment:
nit: you don't need braces for single line block statements (https://flutter.io/style-guide/#avoid-using-braces-for-one-line-long-control-structure-statements)